provider/image.go

```diff
@@ -0,0 +1,314 @@
+package provider
+
+import (
+	"bufio"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/jsonmessage"
+	structpb "github.com/golang/protobuf/ptypes/struct"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+)
+
+const defaultDockerfile = "Dockerfile"
+
+type Image struct {
+	Name     string
+	SkipPush bool
+	Registry Registry
+	Tag      string
+	Build    Build
+}
+
+type Registry struct {
+	Server   string
+	Username string
+	Password string
+}
+
+type Build struct {
+	Context      string
+	Dockerfile   string
+	CachedImages []string
+	Env          map[string]string
+	Args         map[string]*string
+	ExtraOptions []string
+	Target       string
+}
+
+func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
+	urn resource.URN,
+	props *structpb.Struct) (string, *structpb.Struct, error) {
+
+	inputs, err := plugin.UnmarshalProperties(props, plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return "", nil, err
+	}
+
+	//set Registry
+	var reg Registry
+	if !inputs["registry"].IsNull() {
+		reg = Registry{
+			Server:   inputs["registry"].ObjectValue()["server"].StringValue(),
+			Username: inputs["registry"].ObjectValue()["username"].StringValue(),
+			Password: inputs["registry"].ObjectValue()["password"].StringValue(),
+		}
+	}
+
+	// read in values to Image
+	img := Image{
+		Name:     inputs["imageName"].StringValue(),
+		SkipPush: inputs["skipPush"].BoolValue(),
+		Registry: reg,
+	}
+
+	build := marshalBuild(inputs["build"])
+	cache := getCachedImages(img, inputs["build"])
+
+	build.CachedImages = cache
+	img.Build = build
+
+	docker, err := client.NewClientWithOpts(client.FromEnv)
+
+	if err != nil {
+		panic(err)
+	}
+
+	err = p.host.Log(ctx, "info", urn, "Building the image")
+
+	if err != nil {
+		return "", nil, err
+	}
+
+	// make the build context
+	tar, err := archive.TarWithOptions(img.Build.Context, &archive.TarOptions{})
+	if err != nil {
+		return "", nil, err
+	}
+
+	// make the build options
+
+	opts := types.ImageBuildOptions{
+		Dockerfile: img.Build.Dockerfile,
+		Tags:       []string{img.Name}, //this should build the image locally, sans registry info
+		Remove:     true,
+		//CacheFrom:  img.Build.CachedImages, // TODO: this needs a login, so needs to be handled differently.
+		BuildArgs: build.Args,
+		//Version:   types.BuilderBuildKit, // TODO: parse this setting from the `env` input
```

Comment:
Should we open an issue to track how we want to enable build kit? Should this be on as default?